File: 445.两数相加2/addTwoNumbers.go
package problem0445

type ListNode struct {
	Val  int
	Next *ListNode
}

type Stack struct {
	slice []int
}

func (this *Stack) isEmpty() bool {
	return len(this.slice) == 0
}

func (this *Stack) push(val int) {
	this.slice = append(this.slice, val)
}

func (this *Stack) pop() int {
	if this.isEmpty() {
		return 0
	}
	result := this.slice[len(this.slice)-1]
	this.slice = this.slice[:len(this.slice)-1]
	return result
}

// 栈！
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	stack1, stack2 := &Stack{slice: make([]int, 0)}, &Stack{slice: make([]int, 0)}
	for l1 != nil {
		stack1.push(l1.Val)
		l1 = l1.Next
	}
	for l2 != nil {
		stack2.push(l2.Val)
		l2 = l2.Next
	}
	var result *ListNode
	carry := 0
	for !stack1.isEmpty() || !stack2.isEmpty() || carry > 0 {
		carry += stack1.pop() + stack2.pop()

		node := &ListNode{Val: carry % 10, Next: result}
		result = node

		carry /= 10
	}
	return result
}

File: 445.两数相加2/addTwoNumbers.js
// 两数相加 II
// 给你两个 非空 链表来代表两个非负整数。数字最高位位于链表开始位置。它们的每个节点只存储一位数字。将这两数相加会返回一个新的链表。
// 你可以假设除了数字 0 之外，这两个数字都不会以零开头。

// 这道题 高位 位于链表开头，且不能翻转链表，因此就用栈来完成
function addTwoNumbers(l1, l2) {
	const l1Arr = []
	const l2Arr = []
	while (l1) {
		l1Arr.push(l1.val)
		l1 = l1.next
	}
	while (l2) {
		l2Arr.push(l2.val)
		l2 = l2.next
	}
	// 命名为tail的原因是，高位位于链表头
	let tail = null
	let carry = 0
	while(l1Arr.length || l2Arr.length || carry > 0) {
		let sum = carry
		if (l1Arr.length) {
			sum += l1Arr.pop()
		}
		if (l2Arr.length) {
			sum += l2Arr.pop()
		}
		carry = Math.floor(sum / 10)
		tail = { next: tail, val: sum % 10 }
	}
	return tail
}